### PR TITLE
Improve training loop

### PR DESCRIPTION
**Pull request by zparis99**
There's a lot of changes here and there's still more I want to do but I think I should get this out there since it will make follow up changes easier than the current setup.

This includes code for:

- Cleanup
- Breaking up the training loop to make it easier to parse and change. Reduces code copying.
- Adds tests and fixes some bugs.
- Changes how metrics are logged.
- Adds support for tensorboard (see here for how to use: https://researchcomputing.princeton.edu/support/knowledge-base/tensorflow#TensorBoard) Allows for live-updating of plots like this: 
image

- Adds more config options.
- Fixes the lr scheduler not updating.

There are follow up things that still need to be added:
- The correlation code I added sometimes returns NaN, figure out why. 
- The correlation code is pretty inefficient so I only call it during testing. See if there's a way to speed it up.
- Adding more metrics to tensorboard (correlation per electrode, correlation per band, etc).
- Plotting sample reconstructed signals over actual signals for visual comparison in tensorboard.

### CODE REVIEW

File: ECoG_MAE/decoding_main.py
Comment (itamarJa) [inline ECoG_MAE/decoding_main.py:1]: I'm approving the new code but we need to strategize a little here. The current decoding approach isn't very good (averaging)

File: ECoG_MAE/mae_st_util/misc.py
Comment (itamarJa) [inline ECoG_MAE/mae_st_util/misc.py:1]: Making sure - this is an external module imported as-is?

File: submit.sh
Comment (itamarJa) [inline submit.sh:1]: export env yaml and add to repo?

**Review by itamarJa — APPROVED** [review]

File: ECoG_MAE/decoding_main.py
Comment (zparis99) [reply ECoG_MAE/decoding_main.py:1]: Yeah I think there's a few ways we could go about it, if we want to output one decoded word embedding per-2 second sample I believe the usual way to do so is to either average or add a [CLS] token output which we train over for the decoding task.

I think similar to what you were suggesting in your email which could be interesting is increasing the patch size (so more electrodes per patch) and then doing decoding over each of the output embeddings individually. We could then see how the performance performs as a function of time (can map output embeddings back to individual input patches). Each embedding probably incorporates information from the other input patches though so it's maybe not a perfect mapping from the inputs. 

Maybe today at the meeting is a good time to discuss more!

File: ECoG_MAE/mae_st_util/misc.py
Comment (zparis99) [reply ECoG_MAE/mae_st_util/misc.py:1]: Yes, https://github.com/MedARC-AI/fMRI-foundation-model/blob/main/src/mae_utils/misc.py although I only imported the code that I needed as of now for metric logging.

File: submit.sh
Comment (zparis99) [reply submit.sh:1]: Ah good catch I didn't mean to push that. I've changed the submit script to use the virtual environment created when running ./setup.sh so there's no need to export the env yaml.

File: ECoG_MAE/decoding_main.py
Comment (itamarJa) [reply ECoG_MAE/decoding_main.py:1]: In the original decoding task a much shorter period was used. So we would need to either use the positional embeddings to extract signal for relevant frames or else to implement some sort of convolution
